secp256kfun/src/poly.rs

```diff
@@ -0,0 +1,323 @@
+                    }
+                    let higher_degree = s!({ coefficients.last().unwrap() } * a_m);
+                    updated_coefficients.push(higher_degree.public());
+                    coefficients = updated_coefficients;
+                }
+            }
+            coefficients
+        })
+        .collect()
+}
+
+/// Find the coefficients of the polynomial that interpolates a set of points (index, point).
+///
+/// Panics if the indicies are not unique.
+///
+/// A vector with a tail of zero coefficients means the interpolation was overdetermined.
+pub fn interpolate_point_polynomial(
+    points_at_indicies: Vec<(Scalar<Public, impl ZeroChoice>, Point)>,
+) -> Vec<Point<impl PointType, Public, Zero>> {
+    let (indicies, points): (Vec<_>, Vec<_>) = points_at_indicies.into_iter().unzip();
+
+    let basis_polynomials: Vec<_> = unit_basis_polys(indicies.as_slice());
+
+    let interpolating_basis: Vec<_> = basis_polynomials
+        .iter()
+        .zip(points)
+        .map(|(basis_poly, y_value)| point_mul(basis_poly, &y_value))
+        .collect();
+
+    interpolating_basis.into_iter().fold(
+        Vec::with_capacity(basis_polynomials[0].len()),
+        |acc, poly| add(&acc, &poly),
+    )
+}
+
+/// Multiply each coefficient in a scalar polynomial by a point.
+pub fn point_mul(
+    poly: &[Scalar<impl Secrecy, impl ZeroChoice>],
+    point: &Point,
+) -> Vec<Point<NonNormal, Public, Zero>> {
+    poly.iter()
+        .map(|coeff| g!(coeff * point).mark_zero())
+        .collect::<Vec<_>>()
+}
+
+/// Add the coefficients of two point polynomials.
+///
+/// Handles mismatched polynomial lengths.
+pub fn add<T: PointType + Default, S: Secrecy, Z: ZeroChoice>(
+    poly1: &[Point<T, S, Z>],
+    poly2: &[Point<T, S, Z>],
+) -> Vec<Point<NonNormal, Public, Zero>> {
+    let (long, short) = if poly1.len() >= poly2.len() {
+        (poly1, poly2)
+    } else {
+        (poly2, poly1)
+    };
+
+    long.iter()
+        .map(|c| c.mark_zero())
+        .zip(
+            short
+                .iter()
+                .map(|c| c.mark_zero())
+                .chain(iter::repeat(Point::zero())),
+        )
+        .map(|(c1, c2)| g!(c1 + c2))
+        .collect()
+}
+
+/// Interpolate a set of shamir secret shares to find the joint secret.
+///
+/// Each shamir secret share is associated with a participant index (index, share).
+///
+/// Panics if the indicies are not unique.
+pub fn reconstruct_shared_secret(
+    secrets_at_indices: Vec<(Scalar, Scalar<impl Secrecy, impl ZeroChoice>)>,
+) -> Scalar {
+    let (indicies, secrets): (Vec<_>, Vec<_>) = secrets_at_indices.into_iter().unzip();
+    let coefficients: Vec<_> = indicies
+        .iter()
+        .map(|my_index| {
+            lagrange_lambda(
+                *my_index,
+                indicies.clone().into_iter().filter(|j| j != my_index),
+            )
+        })
+        .collect();
+    secrets
```

Comment:
This doesn't need to do two iterations. You can just multiply in the first iteration.